openverse_catalog/dags/providers/provider_api_scripts/finnish_museums.py

```diff
@@ -1,167 +1,128 @@
+    endpoint = "https://api.finna.fi/api/v1/search"
+    batch_limit = 100
+    delay = 5
+    format_type = "0/Image/"
+
+    def ingest_records(self, **kwargs):
+        for building in BUILDINGS:
+            logger.info(f"Obtaining images of building {building}")
+            super().ingest_records(building=building)
+
+    def get_next_query_params(self, prev_query_params, **kwargs):
+        if not prev_query_params:
+            building = kwargs.get("building")
+            return {
+                "filter[]": [f'format:"{self.format_type}"', f'building:"{building}"'],
+                "limit": self.batch_limit,
+                "page": 1,
+            }
+        return {**prev_query_params, "page": prev_query_params["page"] + 1}
+
+    def get_media_type(self, record):
+        return "image"
+
+    def get_batch_data(self, response_json):
+        if (
+            response_json is None
+            or str(response_json.get("status")).lower() != "ok"
+            or response_json.get("records") is None
+            or len(response_json.get("records")) == 0
+        ):
+            return None
+
+        return response_json.get("records")
+
+    def get_record_data(self, data):
+        records = []
+
+        license_url = self.get_license_url(data)
+        if license_url is None:
+            return None
+
+        foreign_identifier = data.get("id")
+        title = data.get("title")
+        building = data.get("buildings")[0].get("value")
+        source = next(
+            (s for s in SUB_PROVIDERS if building in SUB_PROVIDERS[s]), PROVIDER
         )
-    return total_images
+        foreign_landing_url = self._get_landing(data)
+        raw_tags = self._get_raw_tags(data)
+
+        image_list = data.get("images")
+        for img in image_list:
+            image_url = self._get_image_url(img)
+            records.append(
+                {
+                    "license_info": get_license_info(license_url),
+                    "foreign_identifier": foreign_identifier,
+                    "foreign_landing_url": foreign_landing_url,
+                    "image_url": image_url,
+                    "title": title,
+                    "source": source,
+                    "raw_tags": raw_tags,
+                }
+            )
+        return records
+
+    @staticmethod
+    def get_license_url(obj):
+        license_url = obj.get("imageRights", {}).get("link")
+        if license_url is None:
+            return None
+
+        # The API returns urls linking to the Finnish version of the license deed,
+        # (eg `licenses/by/4.0/deed.fi`), but the license validation logic expects
+        # links to the license page (eg `license/by/4.0`).
+        return license_url.removesuffix("deed.fi")
+
+    @staticmethod
+    def _get_raw_tags(obj):
+        tag_lists = obj.get("subjects")
+        if tag_lists is None:
+            return None
+        return list(chain(*tag_lists))
+
+    @staticmethod
```

Comment:
Do we really need to have these two methods separate? I remember reading that Python's function calls can be expensive, and these two functions could be replaced with a couple of lines?

Reply:
It's definitely not _necessary_ to split them, just a slight readability concern to keep `get_record_data` as simple as possible. These are simple enough that I wouldn't be opposed to getting rid of them, though.

Function calls _are_ relatively expensive in Python, but normally I'd consider that a micro-optimization. In the case of Finnish Museums though we've been running up against execution timeouts for awhile now, so we may as well! Good looking out.